### PR TITLE
Dev1.7.0

### PR DESCRIPTION
**Pull request by philjord**
New GL2ES2Pipeline
Readme.md improved
Orginal sun tutorials added
User guide for new pipeline
Many small bug fixes and minor improvements
Note ALL 1.6.0 commits replicated in this branch, not including build.xml changes

### CODE REVIEW

File: src/main/java-templates/org/jogamp/java3d/VersionInfo.java [hunk 1]
```diff
@@ -306,6 +320,14 @@ static String getSpecificationVendor() {
     static String getVersion() {
 	return VERSION;
     }
+    
+    /**
+     * Returns the implementation version string.
+     * @return the implementation version string
+     */
+    static String getVersionFull() {
+	return VERSION;
```
Comment (gouessej) [inline src/main/java-templates/org/jogamp/java3d/VersionInfo.java]: Shouldn't it be VERSION_FULL instead???

File: src/main/java/org/jogamp/java3d/NodeComponentRetained.java [hunk 2]
```diff
@@ -160,9 +155,13 @@ synchronized void notifyUsers() {
 	    return;
 	}
 
-	for (int i=users.size()-1; i >=0; i--) {
-	    ((NodeRetained)users.get(i)).notifySceneGraphChanged(false);
-	}
+	// the reverse order does not appear to be compulsory
+	// cahnge from ArayList to LinkedHashMap prevents ordered traversal
```
Comment (gouessej) [inline src/main/java/org/jogamp/java3d/NodeComponentRetained.java]: "cahnge"??? "change"?

File: src/main/java/org/jogamp/java3d/ShaderAttributeSetRetained.java [hunk 3]
```diff
@@ -313,14 +330,17 @@ synchronized void updateMirrorObject(int component, Object value) {
 	    ShaderAttributeRetained mirrorSA = (ShaderAttributeRetained)value;
  	    assert(mirrorSA != null);
 	    ((ShaderAttributeSetRetained)mirror).attrs.put(mirrorSA.attrName, mirrorSA);
+	    ((ShaderAttributeSetRetained)mirror).attrsValues.add(mirrorSA);
 	}
 	else if((component & ShaderConstants.ATTRIBUTE_SET_REMOVE) != 0) {
 	    // System.err.println("     -- ATTRIBUTE_SET_REMOVE");
```
Comment (gouessej) [inline src/main/java/org/jogamp/java3d/ShaderAttributeSetRetained.java]: Maybe you can remove this commented line of code.

File: src/main/java/org/jogamp/java3d/ShaderAttributeSetRetained.java [hunk 4]
```diff
@@ -313,14 +330,17 @@ synchronized void updateMirrorObject(int component, Object value) {
...
-	    ((ShaderAttributeSetRetained)mirror).attrs.remove((String)value);
+		 ((ShaderAttributeSetRetained)mirror).attrsValues.remove(((ShaderAttributeSetRetained)mirror).attrs.remove((String)value));
+	    
 	}
 	else if((component & ShaderConstants.ATTRIBUTE_SET_CLEAR) != 0) {
 	    // System.err.println("     -- ATTRIBUTE_SET_CLEAR");
```
Comment (gouessej) [inline src/main/java/org/jogamp/java3d/ShaderAttributeSetRetained.java]: Here too...

File: src/main/java/org/jogamp/java3d/Shape3DRetained.java [hunk 5]
```diff
@@ -2374,9 +2375,10 @@ else if(nullCnt == size) {
 	VirtualUniverse firstFndUniv = null;
 
 	synchronized(userList) {
-	    for (int i = userList.size()-1; i >=0; i--) {
-		ms = (Shape3DRetained) userList.get(i);
-
+	    //for (int i = userList.size()-1; i >=0; i--) {
+		//ms = (Shape3DRetained) userList.get(i);
```
Comment (gouessej) [inline src/main/java/org/jogamp/java3d/Shape3DRetained.java]: Maybe you can remove these commented lines of code.

File: src/main/java/org/jogamp/java3d/SparseArray.java [hunk 6]
```diff
@@ -0,0 +1,584 @@
...
+				}
+
+				o++;
+			}
+		}
+
+		mGarbage = false;
+		mSize = o;
+
+		// Log.e("SparseArray", "gc end with " + mSize);
```
Comment (gouessej) [inline src/main/java/org/jogamp/java3d/SparseArray.java]: Maybe you can remove this commented line of code.

**Comment by gouessej** [comment]
I have just posted a few comments. It's mostly ok for me, good job as usual 👍 
Don't forget to update the bug report concerning the enumerations.

**Comment by philjord** [comment]
opps sorry :)

**Comment by gouessej** [comment]
Keep up the good work. I'll have to create some official repositories for Java3D and you'll have to commit to JogAmp GIT repositories too like me (for another engine).

**Pull request merged**